File: src/languages/arabic/diacritizer.ts
/**
 * Neural Arabic diacritizer (Phase 2): a sentence-level BiLSTM that restores short vowels on bare Arabic
 * text, run via ONNX Runtime as an ASYNC PRE-PASS. Its vocalized output feeds the (synchronous) g2p in
 * arabic.ts — so phonemize() stays sync and dependency-free; only this pre-pass touches ONNX. `onnxruntime-
 * node` is an OPTIONAL dependency, imported lazily.
 *
 * Ported from the espeak-ng-portable integration. The shipped .onnx is a permissively-sourced model:
 * CATT teacher (Apache-2.0) → Arabic Wikipedia (CC-BY-SA) silver-only, no NC/GPL. See diacritizer.PROVENANCE.md.
 * Position-preserving: only harakat are inserted after Arabic letters; digits/punctuation/spacing are kept.
 */
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const SHADDA = "ّ";
const LABEL_VOWEL_MARK: Record<string, string> = {
  a: "َ", u: "ُ", i: "ِ", o: "ْ", F: "ً", N: "ٌ", K: "ٍ", "^": "ٰ",
};
const MARK_TO_VOWEL: Record<string, string> = Object.fromEntries(Object.entries(LABEL_VOWEL_MARK).map(([k, v]) => [v, k]));
const IS_MARK = new Set([SHADDA, ...Object.keys(MARK_TO_VOWEL)]);

function isArabicLetterCp(cp: number): boolean {
  return (cp >= 0x0620 && cp <= 0x064a) || (cp >= 0x0671 && cp <= 0x06d3) || cp === 0x0629 || cp === 0x0649;
}
/** Strip every combining haraka → the bare consonant skeleton. */
function undiacritize(word: string): string {
  let out = "";
  for (const c of word) if (!IS_MARK.has(c)) out += c;
  return out;
}
/** Reconstruct the combining marks for one label ("0" = bare). */
function labelToMarks(label: string): string {
  if (label === "0") return "";
  let out = "", v = label;
  if (label.startsWith("~")) { out += SHADDA; v = label.slice(1); }
  if (v && v !== "0") out += LABEL_VOWEL_MARK[v] ?? "";
  return out;
}

// Defective-spelling closed class: high-frequency words whose long /aː/ is an unwritten dagger-alif.
const AR_DEFECTIVE_SPELLING = new Map<string, string>([
  ["هذا", "هٰذَا"], ["هذه", "هٰذِهِ"], ["هذان", "هٰذَان"], ["هذين", "هٰذَيْن"],
  ["ذلك", "ذٰلِكَ"], ["ذلكم", "ذٰلِكُم"], ["كذلك", "كَذٰلِكَ"], ["لكن", "لٰكِن"], ["هؤلاء", "هٰؤُلَاءِ"],
  ["ثلاثمئة", "ثَلَاثُمِئَة"], ["أربعمئة", "أَرْبَعُمِئَة"], ["خمسمئة", "خَمْسُمِئَة"], ["ستمئة", "سِتْتُمِئَة"],
  ["سبعمئة", "سَبْعُمِئَة"], ["ثمانمئة", "ثَمَانِمِئَة"], ["تسعمئة", "تِسْعُمِئَة"],
]);
/** Classical otiose-alif مائة "hundred" → modern مئة (reads /miʔa/, not /maːʔa/). */
function normalizeArabicNumberSpelling(skeleton: string): string {
  return skeleton.replace(/مائة/g, "مئة").replace(/مائت/g, "مئت");
}
/** Rewrite defective-spelling words to their dagger-alif form (keyed on the skeleton; strips a leading proclitic). */
function applyDefectiveSpelling(line: string): string {
  return line.replace(/[؀-ۿݐ-ݿ]+/gu, (w) => {
    const direct = AR_DEFECTIVE_SPELLING.get(undiacritize(w));
    if (direct !== undefined) return direct;
    const m = /^([وفبكل][ً-ْٰ]*)(.+)$/u.exec(w);
    if (m) { const stem = AR_DEFECTIVE_SPELLING.get(undiacritize(m[2]!)); if (stem !== undefined) return m[1]! + stem; }
    return w;
  });
}

/** Whitespace set matching C# char.IsWhiteSpace + Python .split() (NOT JS /\s/) for cross-engine parity. */
function isWhitespace(c: string): boolean {
  const cp = c.charCodeAt(0);
  return cp === 0x20 || (cp >= 0x09 && cp <= 0x0d) || cp === 0x85 || cp === 0xa0 || cp === 0x1680
    || (cp >= 0x2000 && cp <= 0x200a) || cp === 0x2028 || cp === 0x2029 || cp === 0x202f || cp === 0x205f || cp === 0x3000;
}
/** A pausal (phrase-final) boundary follows: end / whitespace / punctuation (not a letter or mark). */
function isPausalBoundary(next: string | undefined): boolean {
  if (next === undefined || isWhitespace(next)) return true;
  const cp = next.codePointAt(0)!;
  const isMark = (cp >= 0x064b && cp <= 0x065f) || cp === 0x0670;
  return !isArabicLetterCp(cp) && !isMark;
}
/** Pausal form for TTS: drop a word-final case-ending vowel/tanwin; accusative ـًا/ـًى → /aː/. */
function pausalize(text: string): string {
  const chars = [...text];
  const out: string[] = [];
  for (let i = 0; i < chars.length; i++) {
    const cp = chars[i]!.codePointAt(0)!;
    if (cp === 0x064b) {
      const nextCp = chars[i + 1]?.codePointAt(0);
      if ((nextCp === 0x0627 || nextCp === 0x0649) && isPausalBoundary(chars[i + 2])) { out.push("َ"); continue; }
    }
    if (cp >= 0x064b && cp <= 0x0650 && isPausalBoundary(chars[i + 1])) continue;
    out.push(chars[i]!);
  }
  return out.join("");
}

export interface DiacritizerMeta { chars: Record<string, number>; labels: Record<string, number>; }
export interface ArabicDiacritizer { diacritize(text: string): Promise<string>; }

interface OrtLike {
  InferenceSession: { create(path: string | Uint8Array): Promise<OrtSession> };
  Tensor: new (type: "int64", data: BigInt64Array, dims: number[]) => unknown;
}
interface OrtSession { run(feeds: Record<string, unknown>): Promise<Record<string, { data: unknown }>>; }

let ortPromise: Promise<OrtLike> | undefined;
async function loadOrt(): Promise<OrtLike> {
  if (!ortPromise) {
    ortPromise = import("onnxruntime-node").then((m) => (m.default ?? m) as unknown as OrtLike).catch(() => {
      throw new Error("Arabic diacritization needs the optional dependency `onnxruntime-node`. Install it with `npm install onnxruntime-node`.");
    });
  }
  return ortPromise;
}

/** Load a neural Arabic diacritizer from ONNX model bytes + sidecar meta. Session created once and reused. */
export async function loadArabicDiacritizer(modelBytes: Uint8Array, meta: DiacritizerMeta): Promise<ArabicDiacritizer> {
  const ort = await loadOrt();
  const session = await ort.InferenceSession.create(modelBytes);
  const ilabels: string[] = [];
  for (const [lab, idx] of Object.entries(meta.labels)) ilabels[idx] = lab;
  const UNK = meta.chars["<unk>"] ?? 1;
  const SP = meta.chars["<sp>"] ?? -1;
  if (SP < 0) throw new Error("diacritizer meta missing the <sp> token");
  const nLabels = ilabels.length;

  async function argmax(indices: number[]): Promise<number[]> {
    const input = new ort.Tensor("int64", BigInt64Array.from(indices, BigInt), [1, indices.length]);
    const out = await session.run({ input });
    const logitsTensor = out["logits"] ?? out[Object.keys(out)[0]!];
    if (!logitsTensor) throw new Error("diacritizer ONNX produced no output");
    const logits = logitsTensor.data as Float32Array;
    const res = new Array<number>(indices.length);
    for (let p = 0; p < indices.length; p++) {
      let best = 0, bestv = -Infinity;
      const base = p * nLabels;
      for (let o = 0; o < nLabels; o++) { const v = logits[base + o]!; if (v > bestv) { bestv = v; best = o; } }
      res[p] = best;
    }
    return res;
  }

  async function diacritizeLine(line: string): Promise<string> {
    const src = [...normalizeArabicNumberSpelling(undiacritize(line))];
    const seq: number[] = [], pos: number[] = [];
    let prevWordHadLetter = false, i = 0;
    while (i < src.length) {
      if (isWhitespace(src[i]!)) { i++; continue; }
      let j = i, hasLetter = false;
      while (j < src.length && !isWhitespace(src[j]!)) { if (isArabicLetterCp(src[j]!.codePointAt(0)!)) hasLetter = true; j++; }
      if (hasLetter) {
        if (prevWordHadLetter) seq.push(SP);
        for (let k = i; k < j; k++) if (isArabicLetterCp(src[k]!.codePointAt(0)!)) { pos.push(k); seq.push(meta.chars[src[k]!] ?? UNK); }
        prevWordHadLetter = true;
      }
      i = j;
    }
    if (seq.length === 0 || seq.every((x) => x === SP)) return line;
    const labelIdx = await argmax(seq);
    const marks = new Map<number, string>();
    let li = 0;
    for (let p = 0; p < seq.length; p++) { if (seq[p] === SP) continue; marks.set(pos[li]!, labelToMarks(ilabels[labelIdx[p]!]!)); li++; }
    return pausalize(applyDefectiveSpelling(src.map((ch, k) => ch + (marks.get(k) ?? "")).join("")));
  }

  return {
    async diacritize(text: string): Promise<string> {
      const lines = text.split("\n");                    // one sentence per line = one inference context
      for (let k = 0; k < lines.length; k++) if (lines[k]!.trim()) lines[k] = await diacritizeLine(lines[k]!);
      return lines.join("\n");
    },
  };
}

/** Load the diacritizer from the model + meta beside this file (model is gitignored — dev stand-in or the
 *  built permissive model). Returns undefined if the .onnx is absent. */
export async function createArabicDiacritizer(): Promise<ArabicDiacritizer | undefined> {
  const dir = dirname(fileURLToPath(import.meta.url));
  let bytes: Buffer;
  try { bytes = readFileSync(join(dir, "diacritizer.onnx")); } catch { return undefined; }
  const meta = JSON.parse(readFileSync(join(dir, "diacritizer.meta.json"), "utf8")) as DiacritizerMeta;
  return loadArabicDiacritizer(new Uint8Array(bytes), meta);
}

File: csharp/Vernacula.Phonemizer/Languages/Arabic/Diacritizer.cs
/**
 * Neural Arabic diacritizer: a sentence-level BiLSTM that restores short vowels on bare Arabic
 * text, run via ONNX Runtime as an ASYNC PRE-PASS. Its vocalized output feeds the (synchronous) g2p in
 * arabic.ts — so phonemize() stays sync and dependency-free; only this pre-pass touches ONNX. `onnxruntime-
 * node` is an OPTIONAL dependency, imported lazily.
 *
 * Position-preserving: only harakat are inserted after Arabic letters; digits/punctuation/spacing are kept.
 */
using System.Text.Json;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Arabic;

public static class Diacritizer
{
    private static string SHADDA => Manifest.MANIFEST.Marks.Shadda;
    private static IReadOnlyDictionary<string, string> LABEL_VOWEL_MARK => Manifest.MANIFEST.Diacritizer.LabelMarks; // label → combining mark
    private static readonly Dictionary<string, string> MARK_TO_VOWEL =
        LABEL_VOWEL_MARK.ToDictionary(kv => kv.Value, kv => kv.Key, StringComparer.Ordinal);
    private static readonly HashSet<string> IS_MARK =
        new(MARK_TO_VOWEL.Keys.Prepend(SHADDA), StringComparer.Ordinal);

    private static bool IsArabicLetterCp(int cp) =>
        (cp >= 0x0620 && cp <= 0x064a) || (cp >= 0x0671 && cp <= 0x06d3) || cp == 0x0629 || cp == 0x0649;

    /** Strip every combining haraka → the bare consonant skeleton. */
    private static string Undiacritize(string word)
    {
        var outp = "";
        foreach (var c in Js.CodePoints(word))
            if (!IS_MARK.Contains(c))
                outp += c;
        return outp;
    }

    /** Reconstruct the combining marks for one label ("0" = bare). */
    private static string LabelToMarks(string label)
    {
        if (label == "0") return "";
        var outp = "";
        var v = label;
        if (label.StartsWith("~", StringComparison.Ordinal)) { outp += SHADDA; v = label[1..]; }
        if (v.Length > 0 && v != "0") outp += LABEL_VOWEL_MARK.GetValueOrDefault(v, "");
        return outp;
    }

    // Defective-spelling closed class (authored data in arabic.jsonc): high-frequency words whose long /aː/ is an
    // unwritten dagger-alif.
    private static readonly IReadOnlyDictionary<string, string> AR_DEFECTIVE_SPELLING =
        Manifest.MANIFEST.Diacritizer.DefectiveSpelling;

    private static readonly JsRe OTIOSE_MIA = JsRegex.Compile("مائة", "g");
    private static readonly JsRe OTIOSE_MIT = JsRegex.Compile("مائت", "g");

    /** Classical otiose-alif مائة "hundred" → modern مئة (reads /miʔa/, not /maːʔa/). */
    private static string NormalizeArabicNumberSpelling(string skeleton) =>
        OTIOSE_MIT.Replace(OTIOSE_MIA.Replace(skeleton, "مئة"), "مئت");

    private static readonly JsRe DEFECTIVE_WORD = JsRegex.Compile("[؀-ۿݐ-ݿ]+", "gu");
    private static readonly JsRe DEFECTIVE_PROCLITIC = JsRegex.Compile("^([وفبكل][ً-ْٰ]*)(.+)$", "u");

    /** Rewrite defective-spelling words to their dagger-alif form (keyed on the skeleton; strips a leading proclitic). */
    private static string ApplyDefectiveSpelling(string line)
    {
        return DEFECTIVE_WORD.Replace(line, w =>
        {
            if (AR_DEFECTIVE_SPELLING.TryGetValue(Undiacritize(w.Value), out var direct)) return direct;
            var m = DEFECTIVE_PROCLITIC.Match(w.Value);
            if (m.Success && AR_DEFECTIVE_SPELLING.TryGetValue(Undiacritize(m.Groups[2].Value), out var stem))
                return m.Groups[1].Value + stem;
            return w.Value;
        });
    }

    /** Whitespace set matching C# char.IsWhiteSpace + Python .split() (NOT JS /\s/) for cross-engine parity. */
    private static bool IsWhitespace(string c)
    {
        var cp = (int)c[0];
        return cp == 0x20 || (cp >= 0x09 && cp <= 0x0d) || cp == 0x85 || cp == 0xa0 || cp == 0x1680
            || (cp >= 0x2000 && cp <= 0x200a) || cp == 0x2028 || cp == 0x2029 || cp == 0x202f || cp == 0x205f || cp == 0x3000;
    }

    /** A pausal (phrase-final) boundary follows: end / whitespace / punctuation (not a letter or mark). */
    private static bool IsPausalBoundary(string? next)
    {
        if (next is null || IsWhitespace(next)) return true;
        var cp = Js.CodePointAt0(next);
        var isMark = (cp >= 0x064b && cp <= 0x065f) || cp == 0x0670;
        return !IsArabicLetterCp(cp) && !isMark;
    }

    /** Pausal form for TTS: drop a word-final case-ending vowel/tanwin; accusative ـًا/ـًى → /aː/.
     *  ⚠ EXPORTED for `tools/arabic/eval_ar_runtime.mts`, which must pausalize the GOLD exactly as the runtime
     *  pausalizes the model or every word-final scores wrong (it read 0% before). Reused rather than ported —
     *  a copy would drift and the measurement would silently stop meaning anything. */
    public static string Pausalize(string text)
    {
        var chars = Js.CodePoints(text);
        var outp = new List<string>();
        for (var i = 0; i < chars.Count; i++)
        {
            var cp = Js.CodePointAt0(chars[i]);
            if (cp == 0x064b)
            {
                int? nextCp = i + 1 < chars.Count ? Js.CodePointAt0(chars[i + 1]) : null;
                if ((nextCp == 0x0627 || nextCp == 0x0649) && IsPausalBoundary(i + 2 < chars.Count ? chars[i + 2] : null))
                {
                    outp.Add("َ");
                    continue;
                }
            }
            if (cp >= 0x064b && cp <= 0x0650 && IsPausalBoundary(i + 1 < chars.Count ? chars[i + 1] : null)) continue;
            outp.Add(chars[i]);
        }
        return string.Concat(outp);
    }

    public sealed class DiacritizerMeta
    {
        public Dictionary<string, int> Chars { get; init; } = new();
        public Dictionary<string, int> Labels { get; init; } = new();
    }

    public interface IArabicDiacritizer
    {
        Task<string> Diacritize(string text);
    }

    private sealed class LoadedDiacritizer : IArabicDiacritizer
    {
        private readonly IOrtLike _ort;
        private readonly IOrtSession _session;
        private readonly DiacritizerMeta _meta;
        private readonly string[] _ilabels;
        private readonly int _unk;
        private readonly int _sp;
        private readonly int _nLabels;

        public LoadedDiacritizer(IOrtLike ort, IOrtSession session, DiacritizerMeta meta)
        {
            _ort = ort;
            _session = session;
            _meta = meta;
            var ilabels = new string[meta.Labels.Count == 0 ? 0 : meta.Labels.Values.Max() + 1];
            foreach (var (lab, idx) in meta.Labels) ilabels[idx] = lab;
            _ilabels = ilabels;
            _unk = meta.Chars.GetValueOrDefault("<unk>", 1);
            _sp = meta.Chars.GetValueOrDefault("<sp>", -1);
            if (_sp < 0) throw new InvalidOperationException("diacritizer meta missing the <sp> token");
            _nLabels = ilabels.Length;
        }

        private async Task<int[]> Argmax(List<int> indices)
        {
            var input = _ort.Tensor("int64", indices.Select(i => (long)i).ToArray(), new[] { 1, indices.Count });
            var outp = await _session.Run(new Dictionary<string, OrtTensor> { ["input"] = input });
            var logitsTensor = outp.GetValueOrDefault("logits") ?? outp.Values.FirstOrDefault();
            if (logitsTensor is null) throw new InvalidOperationException("diacritizer ONNX produced no output");
            var logits = logitsTensor.AsFloat32();
            var res = new int[indices.Count];
            for (var p = 0; p < indices.Count; p++)
            {
                var best = 0;
                var bestv = float.NegativeInfinity;
                var baseIdx = p * _nLabels;
                for (var o = 0; o < _nLabels; o++)
                {
                    var v = logits[baseIdx + o];
                    if (v > bestv) { bestv = v; best = o; }
                }
                res[p] = best;
            }
            return res;
        }

        private async Task<string> DiacritizeLine(string line)
        {
            var src = Js.CodePoints(NormalizeArabicNumberSpelling(Undiacritize(line)));
            var seq = new List<int>();
            var pos = new List<int>();
            var prevWordHadLetter = false;
            var i = 0;
            while (i < src.Count)
            {
                if (IsWhitespace(src[i])) { i++; continue; }
                var j = i;
                var hasLetter = false;
                while (j < src.Count && !IsWhitespace(src[j]))
                {
                    if (IsArabicLetterCp(Js.CodePointAt0(src[j]))) hasLetter = true;
                    j++;
                }
                if (hasLetter)
                {
                    if (prevWordHadLetter) seq.Add(_sp);
                    for (var k = i; k < j; k++)
                        if (IsArabicLetterCp(Js.CodePointAt0(src[k])))
                        {
                            pos.Add(k);
                            seq.Add(_meta.Chars.GetValueOrDefault(src[k], _unk));
                        }
                    prevWordHadLetter = true;
                }
                i = j;
            }
            if (seq.Count == 0 || seq.All(x => x == _sp)) return line;
            var labelIdx = await Argmax(seq);
            var marks = new Dictionary<int, string>();
            var li = 0;
            for (var p = 0; p < seq.Count; p++)
            {
                if (seq[p] == _sp) continue;
                marks[pos[li]] = LabelToMarks(_ilabels[labelIdx[p]]);
                li++;
            }
            var rebuilt = string.Concat(src.Select((ch, k) => ch + marks.GetValueOrDefault(k, "")));
            return Pausalize(ApplyDefectiveSpelling(rebuilt));
        }

        public async Task<string> Diacritize(string text)
        {
            var lines = text.Split('\n');                      // one sentence per line = one inference context
            for (var k = 0; k < lines.Length; k++)
                if (lines[k].Trim().Length > 0)
                    lines[k] = await DiacritizeLine(lines[k]);
            return string.Join("\n", lines);
        }
    }

    /** Load a neural Arabic diacritizer from ONNX model bytes + sidecar meta. Session created once and reused. */
    public static async Task<IArabicDiacritizer> LoadArabicDiacritizer(byte[] modelBytes, DiacritizerMeta meta)
    {
        var ort = await Onnx.LoadOrt("Arabic diacritization");
        var session = await ort.CreateInferenceSession(modelBytes);
        return new LoadedDiacritizer(ort, session, meta);
    }

    /** Load the diacritizer from the model + meta beside this file (model is gitignored — dev stand-in or the
     *  built permissive model). Returns undefined if the .onnx is absent. `variety:"egyptian"` prefers the EGYPTIAN
     *  student model (diacritizer-egy.onnx, restores EGYPTIAN short vowels — مصر→maṣr not the MSA miṣr); it falls
     *  back to the MSA model (→ MSA vowels + the Cairene consonant shifts, the pre-lexicon behavior) if absent. */
    public static async Task<IArabicDiacritizer?> CreateArabicDiacritizer(string? variety = null)
    {
        var bases = variety == "egyptian" ? new[] { "diacritizer-egy", "diacritizer" } : new[] { "diacritizer" };
        foreach (var baseName in bases)
        {
            var modelPath = DataPath.ResolveAllowMissing($"languages/arabic/{baseName}.onnx");
            if (!File.Exists(modelPath)) continue;
            var bytes = File.ReadAllBytes(modelPath);
            var meta = JsonSerializer.Deserialize<DiacritizerMeta>(
                File.ReadAllText(DataPath.ResolveAllowMissing($"languages/arabic/{baseName}.meta.json")),
                Jsonc.JsonOpts)!;
            return await LoadArabicDiacritizer(bytes, meta);
        }
        return null;
    }
}
